hpx/hpx_distributed.cc

```diff
@@ -0,0 +1,369 @@
+            auto &point_deps = deps[point_index];
+            auto &point_rev_deps = rev_deps[point_index];
+
+            // Receive
+            point_n_inputs = 0;
+            if (point >= offset && point < offset + width) {
+              for (auto interval : point_deps) {
+                for (long dep = interval.first; dep <= interval.second; ++dep) {
+                  if (dep < last_offset || dep >= last_offset + last_width) {
+                    continue;
+                  }
+                  // Use shared memory for on-node data.
+                  if (first_point <= dep && dep <= last_point) {
+                    auto output = outputs[dep - first_point];
+                    if (timestep % 2 == 0) {
+                      point_inputs[point_n_inputs].assign(output.begin(),
+                                                        output.end());
+                    } else {
+                      auto output_new = outputs_new[dep - first_point];
+                      point_inputs[point_n_inputs].assign(output_new.begin(),
+                                                        output_new.end());
+                    }
+                  } else {
+                    int from = tag_bits_by_point[dep];
+                    int to = tag_bits_by_point[point];
+                    int tag = (from << 8) | to;
+                    MPI_Request req;
+                    MPI_Irecv(point_inputs[point_n_inputs].data(),
+                              point_inputs[point_n_inputs].size(), MPI_BYTE,
+                              rank_by_point[dep], tag, MPI_COMM_WORLD, &req);
+                    requests.push_back(req);
+                  }
+                  point_n_inputs++;
+                }
+              }
+            }  // receive
+
+            // Send
+            if (point >= last_offset && point < last_offset + last_width) {
+              for (auto interval : point_rev_deps) {
+                for (long dep = interval.first; dep <= interval.second; dep++) {
+                  if (dep < offset || dep >= offset + width ||
+                      (first_point <= dep && dep <= last_point)) {
+                    continue;
+                  }
+                  int from = tag_bits_by_point[point];
+                  int to = tag_bits_by_point[dep];
+                  int tag = (from << 8) | to;
+                  MPI_Request req;
+                  if (timestep % 2 == 0) {
+                      MPI_Isend(point_output.data(), point_output.size(), MPI_BYTE,
+                            rank_by_point[dep], tag, MPI_COMM_WORLD, &req);
+                  } else {
+                      MPI_Isend(point_output_new.data(), point_output_new.size(), MPI_BYTE,
+                            rank_by_point[dep], tag, MPI_COMM_WORLD, &req);
+                  }
+                  requests.push_back(req);
+                }
+              }
+            }  // send
+
+            MPI_Waitall(requests.size(), requests.data(), MPI_STATUSES_IGNORE);
```

Comment:
This is not a critique, but I am curious whether HPX does any integration with MPI here? I.e., does HPX have the ability to swap out another thread while we're waiting for this data to come back?

Reply:
This implementation, we are using plain MPI for inter-node communication.